Create blog article “Defining Network Policy Conformance for Container Network Interface (CNI) providers” (#27159)
kubernetes/website: content/en/blog/_posts/2021-03-09-Defining-NetworkPolicy-Conformance-For-CNIs.md

In addition to cleaning up the validation story for CNI plugins that implement NetworkPolicies,
we've also spent some time improving the Kubernetes NetworkPolicy API for a few commonly requested features.
After months of deliberation, we eventually settled on a few core areas for improvement:

- Port Range policies: We now allow you to specify a *range* of ports for a policy.
  This allows users interested in scenarios like FTP or virtualization to enable advanced policies.
  The port range option for network policies will be available to use in Kubernetes 1.21.
  You can read more about this [here](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2079-network-policy-port-range).
- Namespace as name policies: Allowing users in Kubernetes >= 1.21 to target namespaces using names,
  when building Network Policy objects.  This was done in collaboration with Jordan Liggitt and Tim Hockin on the API Machinery side.
  This change allowed us to improve the Network Policy user experience without actually
  changing the API! For details, you can read about it [here](https://github.com/kubernetes/enhancements/pull/2162).

>   The TLDR is that after Kubernetes 1.21, ALL NAMESPACES will have the following label, added by default:

Reviewer: This has shipped, so I suggest:
```
- Namespace as name policies: Allowing users in Kubernetes ≥ 1.21 to target namespaces using names,
  when building NetworkPolicy objects. This was done in collaboration with Jordan Liggitt and Tim Hockin on the API Machinery side.
  This change allowed us to improve the NetworkPolicy user experience without actually
  changing the NetworkPolicy API! For more details, you can read
  [Automatic labelling](/docs/concepts/overview/working-with-objects/namespaces/#automatic-labelling) in the page about Namespaces.
  The TL;DR: is that for Kubernetes 1.21 and later, **all namespaces** have the following label added by default:
```

Reviewer 2: done 👍 thanks!

    ```

>     kubernetes.io/metadata.name: my-namespace

Reviewer: ```suggestion
    kubernetes.io/metadata.name: <name-of-namespace>
```

    ```

This means you can write a namespace policy against this namespace, even if you can't edit its labels.
For example, this policy, will 'just work', without needing to run a command such as `kubectl edit namespace`.

> In fact, it will even work if you can't edit or view this namespace's data at all, because of the magic of apiserver defaulting.

Reviewer: nit:
```suggestion
In fact, it will even work if you can't edit or view this namespace's data at all, because of the magic of API server defaulting.
```

```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: test-network-policy
  namespace: default
spec:
  podSelector:
    matchLabels:
      role: db
  policyTypes:
  - Ingress

>   ingress:

Reviewer: ```suggestion
  # Allow inbound traffic to Pods labelled role=db, in the namespace 'default'
  # provided that the source is a Pod in the namespace 'my-namespace'
  ingress:
```

  and determine whether *your* specific CNI configuration is fully conformant to the hundreds of different
  Kubernetes Network Policy API constructs.
- Alternatively, you can use a tool like [sonobuoy](https://github.com/vmware-tanzu/sonobuoy)
  to run the existing E2E tests in Kubernetes, with the `--ginkgo.focus=NetworkPolicy` flag.
  Note that only versions of sonobuoy built for k8s 1.21 and above will have the *new* Network Policy tests in them.

## Improvements to the NetworkPolicy API and user experience

In addition to cleaning up the validation story for CNI plugins that implement NetworkPolicies,
we've also spent some time improving the Kubernetes NetworkPolicy API for a few commonly requested features.
After months of deliberation, we eventually settled on a few core areas for improvement:

- Port Range policies: We now allow you to specify a *range* of ports for a policy.
  This allows users interested in scenarios like FTP or virtualization to enable advanced policies.
  The port range option for network policies will be available to use in Kubernetes 1.21.

>   You can read more about this [here](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2079-network-policy-port-range).

Reviewer: ```suggestion
  Read more in [targeting a range of ports](/docs/concepts/services-networking/network-policies/#targeting-a-range-of-ports).
```

Reviewer 2: oops accepting this change cleared the lgtm label 😱
